move saveViewState to address home button case

wagnerch opened the pull request:
This is a possible fix for the issue #401 , apparently when the HOME button is pressed Kodi destroys the window on the C++ side, but there is no callback to Python.  Other things tried in python:

1. override close() method, and call super().close(), but Kodi doesn't fire the Python close.  It just closes in the C++ layer, and probably exits the doModal loop.
2. moving saveViewState outside of exit(), to right after ui.doModal(), but it seems like the controls are all destroyed at this point which is why the view id is saved as empty string.

So what I am doing is whenever the game position changes, doing saveViewState(True), and also whenever the games is rebuilt (due to a filter change) also calling saveViewState.  I haven't really done thorough testing with filters to confirm, but though I would share this with you and it might give you some ideas on a proper fix.

Did test:
1. Putting a search in, moving index to a game, pressing home, restore works.
2. Moving index to a game, pressing home, restore works.
3. Selecting console, press home, restore works.

All of that seems to work, but not confident this is the right way to go about it.  Also the self.gdb.close(), I am assuming that when "del ui" happens in this case that the database is closed via garbage collection, so I left it there.

maloep commented:
I looked into this and I guess I will choose another approach.

I can't reproduce this behaviour to 100%. I am testing on my dev PC where I don't have a remote, only a keyboard. When I test with an Action that executes "ActivateWindow(Home)" it does not save any viewstate at all but it also does not cause any errors on next startup. 

I can only reproduce this behaviour when I try to activate other Widows like TVChannels or Radio that are not completely configured on my system. Kodi then returns to the home screen and opens a dialog with an error message. In this scenario I have issues reopening RCB as described in #401.

In both cases the action id is 122 and the exit method is called. In the first case (Home) the game list is already cleared and RCB returns out of saveViewState without any action. In the second case (TVChannels) RCB goes through the complete code and tries to save the view state. But RCB checks "xbmc.getCondVisibility" of the game list and this is already empty. So it updates the view mode with an empty string and this causes the error on next startup.

My solution would be to update "SETTING_RCB_VIEW_MODE" only when we have a valid view mode id. Alternatively I could hard set the view mode to "50" to make sure the user does not get stuck when he/she always tries to use the home button.

wagnerch commented:
Have you tried this on an GBM-based build of Kodi?  Or is this only Windows where you are testing?  There may be a different behavior with GBM-based builds as it does use a different windowing framework.

Specifically this occurs on Raspberry Pi & Amlogic based devices which are both using GBM.  Upon using the home key (via remote) no action id's are fired, I did throw logging in and the onAction method was never called.

wagnerch commented:
BTW, as mentioned earlier, not sold on this is the right way to solve it.  Just a quick & dirty hack to explore the problem.

I think I explained it differently on the issue ticket, but I think your right because the only way the view mode is corrupted is the onAction method must be called.  But the issue is the controls are already destroyed because the modal window was destroyed by the C++ windowing layer.  So the minimal solution is probably saving the view mode id when it is changed, and carry on with the rest of the view state as is.

I think that's where you were going with this, correct?

wagnerch commented:
Here is another variation on this, but I think the earlier solution was better.  The problem is just the view mode id is saved, but the position in the games list is not, neither would any other controls (search filters) that are read from the controls state.

It seems clear to me that the problem is when ActivateWindow(Home) is fired all of the controls and their states are destroyed before the onAction is called.

So some notes, not sure if the controlId in range... logic is necessary in onFocus, not sure what other controls would gain focus.  But the basic idea is when onFocus on a view control has occurred then save that view state, and represent it with a class variable.

maloep commented:
> 
> 
> Have you tried this on an GBM-based build of Kodi? Or is this only Windows where you are testing? There may be a different behavior with GBM-based builds as it does use a different windowing framework.
> 
> Specifically this occurs on Raspberry Pi & Amlogic based devices which are both using GBM/Linux-based. Upon using the home key (via remote) no action id's are fired, I did throw logging in and the onAction method was never called.

You are right, I forgot to mention that I am only testing on Windows atm. I have some more test devices here but they are not up to date and I don't have time to get them up and running again.

maloep commented:
> 
> 
> BTW, as mentioned earlier, not sold on this is the right way to solve it. Just a quick & dirty hack to explore the problem.
> 
> I think I explained it differently on the issue ticket, but I think your right because the only way the view mode is corrupted is the onAction method must be called. But the issue is the controls are already destroyed because the modal window was destroyed by the C++ windowing layer. So the minimal solution is probably saving the view mode id when it is changed, and carry on with the rest of the view state as is.
> 
> I think that's where you were going with this, correct?

Yes, I think that is the way I want to go. I would save the view mode id whenever it is changed (or could be changed). So, every time showGames is called could be an option. Additionally I would check if we have a valid viewstate before we save it.

What I don't like in your first PR is calling saveViewState everytime the position in the list changes. I am not sure if Kodi fires the event for every single entry but if you are scrolling through large lists with thousands of games I think this could become a performance issue, especially on small devices like RPi.

maloep commented:
> 
> 
> Here is another variation on this, but I think the earlier solution was better. The problem is just the view mode id is saved, but the position in the games list is not, neither would any other controls (search filters) that are read from the controls state.
> 
> It seems clear to me that the problem is when ActivateWindow(Home) is fired all of the controls and their states are destroyed before the onAction is called.
> 
> So some notes, not sure if the controlId in range... logic is necessary in onFocus, not sure what other controls would gain focus. But the basic idea is when onFocus on a view control has occurred then save that view state, and represent it with a class variable.

I think I found the correct place to add this logic. Line 466 to 468 in onClick is the place where the view mode is changed. And showGames is not even called here, so we could miss a change of the view mode if we only store it in showGames.

I still need to play around with it but I think storing the view mode in onClick and on startup I could set it to 50 (GAMES_GROUP_START) if the viewmode is still empty could be the way to go.

wagnerch commented:
> I am not sure if Kodi fires the event for every single entry but if you are scrolling through large lists with thousands of games I think this could become a performance issue, especially on small devices like RPi.

I have around 2000 games on the RPi4 and didn't really notice any performance issue.

wagnerch commented:
> UPDATE: Unfortunately this does not work. The view mode id is the same before and after the change. Looks like Kodi needs some time to update the controls and python does not wait for it.

Yes, that's the same thing I found, I tried onClick.  onFocus fires once you return back to the view, it was the best that I could do.

The other issues other than the view mode are these items are also lost when Home is pressed:
- saveViewMode - control id CONTROL_BUTTON_FAVORITE
- saveViewMode - control id CONTROL_BUTTON_SEARCH
- saveViewState - self.getCurrentListPosition()

I think getCurrentListPosition could be solved by using self.lastPosition, and remove the check on getListSize() and getCurrentListPosition(), and pass it into helper.saveViewState().

wagnerch commented:
added 2 more commits (total of 3) to address the selected game index issues, favorites selected, and search term.

this series of commits fixes the issues with pressing the home button by using class variables to capture the state.  the search term is captured already, added selectedFavorite boolean which is capture in the onClick.

maloep commented:
This looks ok to me. 

One thing we could change: 
In line 700 you are checking if controlId is in GAMES_GROUP_... I think we could check if self.selectedControlId is CONTROL_BUTTON_CHANGE_VIEW. This way we would only update SETTING_RCB_VIEW_MODE when we leave the Change view button.

wagnerch commented:
> One thing we could change:
> In line 700 you are checking if controlId is in GAMES_GROUP_... I think we could check if self.selectedControlId is CONTROL_BUTTON_CHANGE_VIEW. This way we would only update SETTING_RCB_VIEW_MODE when we leave the Change view button.

Oddly that control never fires the onFocus method.  It seems like even the options menu does not fire onFocus either and onFocus is never fired until the user actually changes the view mode and closes the options menu.  Only thing I have seen firing onFocus is game views, didn't check any other dialogs which is why I through in that range check to be safe.

I thought it would be too complex to rely on onClick firing (i.e. setting flag to capture the fact that they clicked CONTROL_BUTTON_CHANGE_VIEW) and then waiting for the onFocus to fire to interpret and save the view mode.  I think as you also found the onClick has a delay until the view is visible, and using onFocus was a relatively clean way to know the view was visible based on the assumption that if it was focused it must be visible.

maloep commented:
> 
> 
> > One thing we could change:
> > In line 700 you are checking if controlId is in GAMES_GROUP_... I think we could check if self.selectedControlId is CONTROL_BUTTON_CHANGE_VIEW. This way we would only update SETTING_RCB_VIEW_MODE when we leave the Change view button.
> 
> EDIT: actually stand corrected. onFocus does fire for CONTROL_BUTTON_CHANGE_VIEW. not sure why i missed that my first time testing it.
> 
There have been errors in the logging functions. I fixed this with my latest commit. Maybe you made your tests before I made this commit.

> So the issue I see here is self.selectedControlId may not be CONTROL_BUTTON_CHANGE_VIEW when onFocus fires the 2nd time, for example if I go into the options menu, change the view, scroll down (control CONTROL_BUTTON_FAVORITE is now focused), and then leave the options menu the self.selectedControlId (last control focused) would be CONTROL_BUTTON_FAVORITE and the controlId fired for onFocus would be the game view mode.
> 
We can check IF self.selectedControlId is CONTROL_BUTTON_CHANGE_VIEW and only do our update when this is the case. At least on my windows machine this works as I expect it. So the second call would be ignored if we do it like this.

> One optimization is maybe checking self.selectedViewModeId changed, or perhaps deferring the setSetting call back to the saveViewMode and have it reference self.selectedViewModeId instead of accessing the control. I am not sure if you looking to minimize updates to the settings file, or what your thoughts are there?

Yes, I want to minimize updates to the settings file and unnecessary calls in general (I know RCB is full of it, but we have to start somewhere:)).

I could merge your PR and do the last bits as I see fit if you are ok with it.

wagnerch commented:
> We can check IF self.selectedControlId is CONTROL_BUTTON_CHANGE_VIEW and only do our update when this is the case. At least on my windows machine this works as I expect it. So the second call would be ignored if we do it like this.

I think you will have to test it out, but basically this is the scenario on GBM/Linux and I suspect Windows will be the same:

1. Up button on remote to bring up options
2. onFocus fires: CONTROL_BUTTON_CHANGE_VIEW
3. Click on "change view"
4. onClick fires: CONTROL_BUTTON_CHANGE_VIEW
5. Down button on remote
6. onFocus fires: CONTROL_BUTTON_FAVORITE
7. Right button on remote
8. onFocus fires: one of GAME_LISTS

So in this situation, if your relying on onFocus alone it wouldn't work as self.lastControlId would be CONTROL_BUTTON_FAVORITE.  onFocus does not fire on the view control until you close the options menu.

I guess you could use a flag in step 4 to indicate they pushed the button to change the view, but you still have to wait for the onFocus since there is a delay between click on change view, Container.NextViewMode, and the actual control becoming focused/visible.  I may not follow what your saying, but maybe give this test case a shot with the solution you work out and see if it works as expected.

The last force-push moved the settings update into saveViewMode where it was originally, and I just set the self.selectedViewModeId in the onFocus method.  I agree with you, probably a good idea to minimize updating settings file, reduces the risk of corrupting the file and minimizes wear on SD cards & SSDs.  And it looks like using class variables is a good way to mitigate this problem, just figuring out the right place to capture the various control states is all we need to sort out.

So far these three fixes seem to resolve all of the setting issues when the Home button is used.

maloep commented:
> So in this situation, if your relying on onFocus alone it wouldn't work as self.lastControlId would be CONTROL_BUTTON_FAVORITE. onFocus does not fire on the view control until you close the options menu.

I misread your code. I did not realize that you store controlId in viewmodeId. So it makes sense to check if we have a GAMESGROUP control.

I still have two issues with the current approach (I tried to comment directly in your commit but somehow I did not manage to see all your changes to master at once):
- You only store selectedViewModeId when setting SHOWNAVIGATIONHINT is set to true. With SHOWNAVIGATIONHINT set to false RCB won't save the view mode at all.
- You only call saveViewState when you change position in game list. When you change the view and then exit RCB without moving inside the game list, RCB also won't save the view mode.

wagnerch commented:
> I still have two issues with the current approach (I tried to comment directly in your commit but somehow I did not manage to see all your changes to master at once):
> 
>     * You only store selectedViewModeId when setting SHOWNAVIGATIONHINT is set to true. With SHOWNAVIGATIONHINT set to false RCB won't save the view mode at all.
> 
>     * You only call saveViewState when you change position in game list. When you change the view and then exit RCB without moving inside the game list, RCB also won't save the view mode.

I think you might need to look at the latest commits, I force-pushed the branch and revised the commits.  If you click on the Commits tab at the top of the PR it should show the latest versions of the commits.

maloep commented:
I am quite confident that I caught all commits and tested them locally. (I am still not 100% familiar with git and github, so the issue might be on my end.)

Please try the following and check if it is working for you:
1
- Disable RCB Setting SHOWNAVIGATIONHINT
- Open RCB and change the view mode
- Move around in the game list
- Close and Reopen RCB. On my system you will still have the old view mode.

2
- Open RCB
- Change the view mode
- Go back to the games list
- Close RCB without navigating inside the games list. On my system you will still have the old view mode.

wagnerch commented:
Not sure what happened on the PR, but it looks like you collected an old version of the commit 11b7541f1825897e8915a2eb6b888a732b4d966f, that wasn't an active commit on the PR yesterday.  It should be discarded.

In any case, I see what your talking about, however that is how the hold code functioned.  Based on what I see it doesn't actually affect saving of the view mode.  The only logic there is if self.selectedViewModeId is None or self.selectedViewModeId = "" then it will default to 50 if SETTING_RCB_SHOWNAVIGATIONHINT is true.  I am actually thinking those 2 lines could be dropped, and perhaps just change the default value in the class to CONTROL_GAMES_GROUP_START (50), that should cover the new user case (where view mode id is not setup).

That whole section of code all it seems to do is write the message line depending on whether it is vertical or horizontal layout to indicate the cursor/d-pad functions.

I can rework those commits again, but it would basically be to drop these 2 lines:
            if self.selectedViewModeId is None or self.selectedViewModeId == "":
                self.selectedViewModeId = CONTROL_GAMES_GROUP_START

and change this line in the class initialziation:

    selectedViewModeId = None

to

    selectedViewModeId = CONTROL_GAMES_GROUP_START


This latter piece is probably not technically required, because what should happen is when we try to loadSettings and view_mode is empty/non-existent then selectedViewModeId does nothing, and the regular UI will focus on 50 and at that point onFocus fires and sets selectedViewModeId.  At least that's what I am expecting, would have to wipe settings to confirm.  I know I set view_mode to empty and that worked.

wagnerch commented:
did that rework i mentioned earlier, and re-pushed the branch.  take a look at the three commits above.  they are basically the same with the changes in my last comment.

maloep commented:
Sorry for the hassle. I am using GitHub Desktop and somehow it always includes 11b7541 and wants me to merge it into the PR when I try to pull your latest commits . I now made the changes manually. I hope I did not break anything again with this move.

wagnerch commented:
Yeah, it looks like it pulled that commit back in, you might want to rebase instead of a merge, or setup a different branch?  I have never used GitHub desktop, I just use git.  But it's actually putting those commits on my branch because I have the flag set to allow it from a PR.

What I usually do is use the gh cli to pull an active PR, but you can also add my repo on github as an remote, and checkout fix-viewstate to a branch in your local repository.  But not really familiar with GitHub Desktop.  I can re push the branch and reset the reference.

maloep commented:
I think, I got it. I had to undo some local commits to your PR on my system.

Everything worked well on my system with one last issue:
When I use the home button to quit RCB, saveViewState exits with self.getListSize() == 0. But we can just remove this check. I think it is not needed anymore.

wagnerch commented:
> I think, I got it. I had to undo some local commits to your PR on my system.
> 
> Everything worked well on my system with one last issue:
> When I use the home button to quit RCB, saveViewState exits with self.getListSize() == 0. But we can just remove this check. I think it is not needed anymore.

OK, can do that.  I was thinking the same thing, but for some reason it works on Linux.  So I wondering if we leave that to timing or windowing framework differences.

maloep commented:
Finally. Sorry for the long process.
